### DISCUSSION TITLE
PNG to multipage PDF: Is it possible to get different margins for odd and even pages?

### DISCUSSION THREAD

**Discussion by johnmapeson** [ImageMagick/ImageMagick#7572] [Help]
### ImageMagick version

 7.1.1-36

### Operating system, version and so on

macOS Sonoma 14.5

### Description

The following script is to convert a tall screenshot (e.g., a 5000 x 50000 px screenshot of a webpage) to a multipage PDF.

``` sh
#!/bin/zsh
inputimage=input.png
outputpdf=output.pdf
paperwidth=8.5
paperheight=11
marginsleftright=1
marginstopbottom=1
allowupscaling=false
papercolor=white

marginsleftright=${$(( marginsleftright*300 ))%.}
marginstopbottom=${$(( marginstopbottom*300 ))%.}
columnwidth=${$(( paperwidth*300 - marginsleftright*2 ))%.}
columnheight=${$(( paperheight*300 - marginstopbottom*2 ))%.}
if [[ $allowupscaling == 'false' ]]; then
  resizecolumnwidth=${columnwidth}\>
else
  resizecolumnwidth=$columnwidth
fi

magick $inputimage +repage -resize ${resizecolumnwidth}x \
  -crop x${columnheight} +repage \
  -background ${papercolor} -extent ${columnwidth}x${columnheight} \
  -bordercolor ${papercolor} -border ${marginsleftright}x${marginstopbottom} \
  -units PixelsPerInch -density 300 $outputpdf
```

Test image: https://github.com/jsx97/test/blob/main/appleseed.png

It works, but I don't like that if I view two pages side-by-side, the margin between them is too large.

```
Fig. 1
+--------------+--------------+
|MARGINS       |MARGINS       |
|   +------+   |   +------+   |
| 1 |TEXT  |   2   |TEXT  | 1 |
|<->|      |<----->|      |<->|
|   |      |   |   |      |   |
|   |      |   |   |      |   |
|   +------+   |   +------+   |
|              |              |
+--------------+--------------+
```

I would prefer it to look like this:

```
Fig. 2
+------------+------------+
|            |            |
|   +------+ | +------+   |
| 1 |      | 1 |      | 1 |
|<->|      |<->|      |<->|
|   |      | | |      |   |
|   |      | | |      |   |
|   +------+ | +------+   |
|            |            |
+------------+------------+
```

That is, odd pages should be like

```
+------------+
|            |
|   +------+ |
|   |      | |
|   |      | |
|   |      | |
|   |      | |
|   +------+ |
|            |
+------------+
```

and even pages should be like

```
+------------+
|            |
| +------+   |
| |      |   |
| |      |   |
| |      |   |
| |      |   |
| +------+   |
|            |
+------------+
```

Is it possible with ImageMagick? That is, how to change the script I have posted to have the desired margins? Or maybe it would be simpler to post-process PDF using another ImageMagick script?

**Comment by snibgo** [comment]
There are many different ways to do this. One way is to use `-extent` and `-border` to make a wide margin on both sides, and then use `-crop` to remove from either the left or the right.

In `-crop WxH+X+Y` each of WHXY is a number or a `%[fx:...]` expression.

Within an `%[fx:...]` expression, `t` is the number of the image within the list: 0, 1, 2, etc. Within expressions, `%` is the modulus operator. So %[fx:t%2] is 0 or 1.

For example, `rose:` is 70x46. We duplicate it to make multiple images. To each image we add a 20 pixel border, so each image is 110x86 pixels. Then we crop each image to 100x86, with an offset that is either 0 or 10.

```
magick rose: -duplicate 2 -bordercolor Blue -border 20x20 -crop 100x86+%[fx:(t%2)*10]+0 +repage x.png
```


x-0.png:

![x-0](https://github.com/user-attachments/assets/afe8acca-ccbf-495d-9843-54a1155c28fe)

x-1.png:

![x-1](https://github.com/user-attachments/assets/702e1c44-480d-4166-9846-5ed50b01ca33)

x-2.png:

![x-2](https://github.com/user-attachments/assets/34824f9f-3712-4057-9586-d54e7ddd8610)

**Comment by johnmapeson** [comment]
@snibgo Thank you very much! I understand the solution you have posted does work, but maybe you can make a review of my code and tell what's wrong in my current implementation? (This will save me 1-2 days.)

```
#!/bin/zsh
inputimage=appleseed.png
outputpdf=output.pdf
paperwidth=5.5
paperheight=8.5
marginsleftright=1
marginstopbottom=1
allowupscaling=true
papercolor=white

# 8.5 x 11   ANSI A, aka Letter
# 5.5 x 8.5  Half Letter
# 9   x 12   Arch A
# 6   x 9    US Trade

marginsleftright=${$(( marginsleftright*300 ))%.}
marginstopbottom=${$(( marginstopbottom*300 ))%.}
paperwidth=${$(( paperwidth*300 ))%.}
paperheight=${$(( paperheight*300 ))%.}
columnwidth=${$(( paperwidth - marginsleftright*2 ))%.}
columnheight=${$(( paperheight - marginstopbottom*2 ))%.}
if [[ $allowupscaling == 'false' ]]; then
  resizetocolumnwidth=${columnwidth}\>
else
  resizetocolumnwidth=$columnwidth
fi
columnoffset=$(( marginsleftright/2 ))

magick $inputimage +repage -resize ${resizetocolumnwidth}x \
  -crop x${columnheight} +repage \
  -background ${papercolor} -extent ${columnwidth}x${columnheight} \
  -bordercolor ${papercolor} -border ${marginsleftright}x${marginstopbottom} \
  -crop "${paperwidth}x${paperheight}+%[fx:(t%2)*${columnoffset}]+0" +repage \
  -units PixelsPerInch -density 300 $outputpdf
```

I don't see the inner margin has been reduced. Also, now the pages are of the different sizes... :-(

**Comment by snibgo** [comment]
Your original script gives errors for me (Cygwin bash on Windows 11). Perhaps zsh is different.

For example:
```
marginsleftright=${$(( marginsleftright*300 ))%.}
appleseed.sh: line 11: ${$(( marginsleftright*300 ))%.}: bad substitution
```
I fix these by removing `${` and `%.}` like this:
```
marginsleftright=$(( marginsleftright*300 ))
```

Your paperwidth of 8.5 causes problems. `man bash` says for me:
> Evaluation is done in fixed-width integers ...
Integers, not floating-point.

... so:
```
paperwidth=$(( 8.5*300 ))
```
... is 2400, not 2550 as we would hope. I solve that by using awk.

The method I showed above makes the images over-wide, then `-crop` removes columns from either the left or right of each. Your code doesn't make the images over-wide. I fix this with `extleftright`.

I also echo the variables to aid debugging. Bash script (not zsh):
```
#!/bin/bash
inputimage=appleseed.png
outputpdf=output.pdf
paperwidth=8.5
paperheight=11
marginsleftright=1
marginstopbottom=1
allowupscaling=false
papercolor=white

marginsleftright=$(( marginsleftright*300 ))
marginstopbottom=$(( marginstopbottom*300 ))
paperwidth=$(awk "BEGIN { print ${paperwidth}*300 }" )
paperheight=$(( paperheight*300 ))
columnwidth=$(( paperwidth - marginsleftright*2 ))
columnheight=$(( paperheight - marginstopbottom*2 ))
if [[ $allowupscaling == 'false' ]]; then
  resizecolumnwidth=${columnwidth}\>
else
  resizecolumnwidth=${columnwidth}
fi
columnoffset=$(( marginsleftright/2 ))

extleftright=$(( ${columnwidth}+${columnoffset} ))

echo marginsleftright = ${marginsleftright}
echo marginstopbottom = ${marginstopbottom}
echo paperwidth = ${paperwidth}
echo paperheight = ${paperheight}
echo columnwidth = ${columnwidth}
echo columnheight = ${columnheight}
echo resizecolumnwidth = ${resizecolumnwidth}
echo columnoffset = ${columnoffset}
echo extleftright = ${extleftright}

magick.exe $inputimage +repage \
  -resize ${resizecolumnwidth}x \
  -crop x${columnheight} +repage \
  -background ${papercolor} -extent ${extleftright}x${columnheight} \
  -bordercolor ${papercolor} -border ${marginsleftright}x${marginstopbottom} \
  -units PixelsPerInch -density 300 \
  -crop "${paperwidth}x${paperheight}+%[fx:(t%2)*${columnoffset}]+0" +repage \
  $outputpdf
```

**Comment by snibgo** [comment]
I don't understand why you have both `-extent` and `-border`, so here is a version without `-border`. I have made papercolor blue to more easily see what happens.
```
#!/bin/bash
inputimage=appleseed.png
outputpdf=output.pdf
paperwidth=8.5
paperheight=11
marginsleftright=1
marginstopbottom=1
allowupscaling=false
papercolor=blue

marginsleftright=$(( marginsleftright*300 ))
marginstopbottom=$(( marginstopbottom*300 ))
paperwidth=$(awk "BEGIN { print ${paperwidth}*300 }" )
paperheight=$(( paperheight*300 ))
columnwidth=$(( paperwidth - marginsleftright*3/2 ))
columnheight=$(( paperheight - marginstopbottom*2 ))
if [[ $allowupscaling == 'false' ]]; then
  resizecolumnwidth=${columnwidth}\>
else
  resizecolumnwidth=${columnwidth}
fi
columnoffset=$(( marginsleftright/2 ))

extleftright=$(( columnwidth+columnoffset*4 ))
exttopbottom=$(( paperheight ))

echo marginsleftright = ${marginsleftright}
echo marginstopbottom = ${marginstopbottom}
echo paperwidth = ${paperwidth}
echo paperheight = ${paperheight}
echo columnwidth = ${columnwidth}
echo columnheight = ${columnheight}
echo resizecolumnwidth = ${resizecolumnwidth}
echo columnoffset = ${columnoffset}
echo extleftright = ${extleftright}
echo exttopbottom = ${exttopbottom}

magick.exe $inputimage +repage \
  -resize ${resizecolumnwidth}x \
  -crop x${columnheight} +repage \
  -gravity Center \
  -background ${papercolor} -extent ${extleftright}x${exttopbottom} \
  -units PixelsPerInch -density 300 \
  +gravity \
  -crop "${paperwidth}x${paperheight}+%[fx:(t%2)*${columnoffset}]+0" +write info: +repage \
  $outputpdf
```

**Comment by johnmapeson** [comment]
> I don't understand why you have both `-extent` and `-border`

I use `-extent` so that the last page will be the same height as others.

- With `-extent`, all the pages are 8.5 x 11 inches
- Without it, the last page is 8.5 x 9.48

I use borders to add page margins. Probably you mean that I can simulate borders using `-extent`, but honestly I don't think this is a good idea. I would prefer to keep these tasks separated.

Sadly, neither solution doesn't work really fine yet.

My original version:

<img width="456" alt="image" src="https://github.com/user-attachments/assets/2b87b0bc-f461-4251-9e6f-9c76039e41e5">

It works fine except that if I view 2 pages side by side, the margin between them is too large.

Your first version:

<img width="458" alt="image" src="https://github.com/user-attachments/assets/1d635c1a-e74b-4abf-8302-17075676bb26">

There, the issue is that it "moves" text blocks of even pages only, but it doesn't "move" text blocks of odd pages.

Your second version:

<img width="576" alt="image" src="https://github.com/user-attachments/assets/5446b3f3-3aaa-4632-8107-fdabfb7da176">

There, there is an obvious problem with the last page, and also I would prefer to not "merge" the task of adding borders with the task of extending the page.
